// @taqueria/plugin-ligo/index.mjs
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// index.ts
import { configurePlugin } from "@taqueria/lib-ligo";

// ligo_templates.ts
var ligo_templates_exports = {};
__export(ligo_templates_exports, {
  jsligo_template: () => jsligo_template,
  mligo_template: () => mligo_template
});
var mligo_template = `
module Counter = struct
  type storage = int

  type ret = operation list * storage

  (* Three entrypoints *)

  [@entry]
  let increment (delta : int) (store : storage) : ret = [], store + delta

  [@entry]
  let decrement (delta : int) (store : storage) : ret = [], store - delta

  [@entry]
  let reset (() : unit) (_ : storage) : ret = [], 0

end
`;
var jsligo_template = `
export namespace Counter {
    export type storage = int;
    type ret = [list<operation>, storage];
    // Three entrypoints
  
    @entry
    const increment = (delta: int, store: storage): ret =>
      [list([]), store + delta];
    @entry
    const decrement = (delta: int, store: storage): ret =>
      [list([]), store - delta];
    @entry
    const reset = (_p: unit, _s: storage): ret => [list([]), 0]
  };
`;

// index.ts
configurePlugin({
  name: "@taqueria/plugin-ligo",
  alias: "ligo",
  dockerImage: "ligolang/ligo:1.2.0",
  dockerImageEnvVar: "TAQ_LIGO_IMAGE",
  canUseLIGOBinary: true,
  unparsedArgs: process.argv,
  templates: ligo_templates_exports
});
