// @dipdup/tzkt-api/dist/lib/services/BigMapsService.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { request as __request } from '../core/request';
var BigMapsService = /** @class */ (function () {
    /**
     * creates a new service
     * @param config the configuration to use
     */
    function BigMapsService(config) {
        this.config = config;
    }
    /**
     * Get bigmaps count
     * Returns the total number of bigmaps.
     * @returns number
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapsCount = function () {
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options = this.getBigMapsCountApiRequestOptions();
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmaps count
     * Returns the total number of bigmaps.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapsCountApiRequestOptions = function () {
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/count" });
    };
    /**
     * Get bigmaps
     * Returns a list of bigmaps.
     * @returns BigMap
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMaps = function (_a) {
        var micheline = _a.micheline, contract = _a.contract, path = _a.path, tags = _a.tags, active = _a.active, lastLevel = _a.lastLevel, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        options = this.getBigMapsApiRequestOptions({
                            micheline: micheline,
                            contract: contract,
                            path: path,
                            tags: tags,
                            active: active,
                            lastLevel: lastLevel,
                            select: select,
                            sort: sort,
                            offset: offset,
                            limit: limit,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmaps
     * Returns a list of bigmaps.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapsApiRequestOptions = function (_a) {
        var micheline = _a.micheline, contract = _a.contract, path = _a.path, tags = _a.tags, active = _a.active, lastLevel = _a.lastLevel, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps", query: {
                'micheline': micheline,
                'contract': contract,
                'path': path,
                'tags': tags,
                'active': active,
                'lastLevel': lastLevel,
                'select': select,
                'sort': sort,
                'offset': offset,
                'limit': limit,
            } });
    };
    /**
     * Get bigmap updates
     * Returns a list of all bigmap updates.
     * @returns BigMapUpdate
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapUpdates = function (_a) {
        var micheline = _a.micheline, bigmap = _a.bigmap, path = _a.path, contract = _a.contract, tags = _a.tags, action = _a.action, value = _a.value, level = _a.level, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        options = this.getBigMapUpdatesApiRequestOptions({
                            micheline: micheline,
                            bigmap: bigmap,
                            path: path,
                            contract: contract,
                            tags: tags,
                            action: action,
                            value: value,
                            level: level,
                            sort: sort,
                            offset: offset,
                            limit: limit,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmap updates
     * Returns a list of all bigmap updates.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapUpdatesApiRequestOptions = function (_a) {
        var micheline = _a.micheline, bigmap = _a.bigmap, path = _a.path, contract = _a.contract, tags = _a.tags, action = _a.action, value = _a.value, level = _a.level, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/updates", query: {
                'micheline': micheline,
                'bigmap': bigmap,
                'path': path,
                'contract': contract,
                'tags': tags,
                'action': action,
                'value': value,
                'level': level,
                'sort': sort,
                'offset': offset,
                'limit': limit,
            } });
    };
    /**
     * Get bigmap by Id
     * Returns a bigmap with the specified Id.
     * @returns BigMap
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapById = function (_a) {
        var id = _a.id, micheline = _a.micheline;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = this.getBigMapByIdApiRequestOptions({
                            id: id,
                            micheline: micheline,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmap by Id
     * Returns a bigmap with the specified Id.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapByIdApiRequestOptions = function (_a) {
        var id = _a.id, micheline = _a.micheline;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id, query: {
                'micheline': micheline,
            } });
    };
    /**
     * Get bigmap type
     * Returns a type of the bigmap with the specified Id in Micheline format (with annotations).
     * @returns MichelinePrim
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapType = function (_a) {
        var id = _a.id;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = this.getBigMapTypeApiRequestOptions({
                            id: id,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmap type
     * Returns a type of the bigmap with the specified Id in Micheline format (with annotations).
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getBigMapTypeApiRequestOptions = function (_a) {
        var id = _a.id;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id + "/type" });
    };
    /**
     * Get bigmap keys
     * Returns a list of bigmap keys.
     * @returns BigMapKey
     * @throws ApiError
     */
    BigMapsService.prototype.getKeys = function (_a) {
        var id = _a.id, micheline = _a.micheline, active = _a.active, key = _a.key, value = _a.value, lastLevel = _a.lastLevel, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        options = this.getKeysApiRequestOptions({
                            id: id,
                            micheline: micheline,
                            active: active,
                            key: key,
                            value: value,
                            lastLevel: lastLevel,
                            select: select,
                            sort: sort,
                            offset: offset,
                            limit: limit,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmap keys
     * Returns a list of bigmap keys.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getKeysApiRequestOptions = function (_a) {
        var id = _a.id, micheline = _a.micheline, active = _a.active, key = _a.key, value = _a.value, lastLevel = _a.lastLevel, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id + "/keys", query: {
                'micheline': micheline,
                'active': active,
                'key': key,
                'value': value,
                'lastLevel': lastLevel,
                'select': select,
                'sort': sort,
                'offset': offset,
                'limit': limit,
            } });
    };
    /**
     * Get bigmap key
     * Returns the specified bigmap key.
     * @returns BigMapKey
     * @throws ApiError
     */
    BigMapsService.prototype.getKey = function (_a) {
        var id = _a.id, key = _a.key, micheline = _a.micheline;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = this.getKeyApiRequestOptions({
                            id: id,
                            key: key,
                            micheline: micheline,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmap key
     * Returns the specified bigmap key.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getKeyApiRequestOptions = function (_a) {
        var id = _a.id, key = _a.key, micheline = _a.micheline;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id + "/keys/" + key, query: {
                'micheline': micheline,
            } });
    };
    /**
     * Get bigmap key updates
     * Returns updates history for the specified bigmap key.
     * @returns BigMapKeyUpdate
     * @throws ApiError
     */
    BigMapsService.prototype.getKeyUpdates = function (_a) {
        var id = _a.id, key = _a.key, micheline = _a.micheline, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        options = this.getKeyUpdatesApiRequestOptions({
                            id: id,
                            key: key,
                            micheline: micheline,
                            sort: sort,
                            offset: offset,
                            limit: limit,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get bigmap key updates
     * Returns updates history for the specified bigmap key.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getKeyUpdatesApiRequestOptions = function (_a) {
        var id = _a.id, key = _a.key, micheline = _a.micheline, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id + "/keys/" + key + "/updates", query: {
                'micheline': micheline,
                'sort': sort,
                'offset': offset,
                'limit': limit,
            } });
    };
    /**
     * Get historical keys
     * Returns a list of bigmap keys at the specific block.
     * @returns BigMapKeyHistorical
     * @throws ApiError
     */
    BigMapsService.prototype.getHistoricalKeys = function (_a) {
        var id = _a.id, level = _a.level, micheline = _a.micheline, active = _a.active, key = _a.key, value = _a.value, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        options = this.getHistoricalKeysApiRequestOptions({
                            id: id,
                            level: level,
                            micheline: micheline,
                            active: active,
                            key: key,
                            value: value,
                            select: select,
                            sort: sort,
                            offset: offset,
                            limit: limit,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get historical keys
     * Returns a list of bigmap keys at the specific block.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getHistoricalKeysApiRequestOptions = function (_a) {
        var id = _a.id, level = _a.level, micheline = _a.micheline, active = _a.active, key = _a.key, value = _a.value, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id + "/historical_keys/" + level, query: {
                'micheline': micheline,
                'active': active,
                'key': key,
                'value': value,
                'select': select,
                'sort': sort,
                'offset': offset,
                'limit': limit,
            } });
    };
    /**
     * Get historical key
     * Returns the specified bigmap key at the specific block.
     * @returns BigMapKeyHistorical
     * @throws ApiError
     */
    BigMapsService.prototype.getKey2 = function (_a) {
        var id = _a.id, level = _a.level, key = _a.key, micheline = _a.micheline;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = this.getKey2ApiRequestOptions({
                            id: id,
                            level: level,
                            key: key,
                            micheline: micheline,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get historical key
     * Returns the specified bigmap key at the specific block.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    BigMapsService.prototype.getKey2ApiRequestOptions = function (_a) {
        var id = _a.id, level = _a.level, key = _a.key, micheline = _a.micheline;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/bigmaps/" + id + "/historical_keys/" + level + "/" + key, query: {
                'micheline': micheline,
            } });
    };
    return BigMapsService;
}());
export { BigMapsService };
