// @dipdup/tzkt-api/dist/lib/services/RightsService.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { request as __request } from '../core/request';
var RightsService = /** @class */ (function () {
    /**
     * creates a new service
     * @param config the configuration to use
     */
    function RightsService(config) {
        this.config = config;
    }
    /**
     * Get rights count
     * Returns the total number of stored rights.
     * @returns number
     * @throws ApiError
     */
    RightsService.prototype.getCount = function (_a) {
        var type = _a.type, baker = _a.baker, cycle = _a.cycle, level = _a.level, slots = _a.slots, priority = _a.priority, status = _a.status;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        options = this.getCountApiRequestOptions({
                            type: type,
                            baker: baker,
                            cycle: cycle,
                            level: level,
                            slots: slots,
                            priority: priority,
                            status: status,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get rights count
     * Returns the total number of stored rights.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    RightsService.prototype.getCountApiRequestOptions = function (_a) {
        var type = _a.type, baker = _a.baker, cycle = _a.cycle, level = _a.level, slots = _a.slots, priority = _a.priority, status = _a.status;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/rights/count", query: {
                'type': type,
                'baker': baker,
                'cycle': cycle,
                'level': level,
                'slots': slots,
                'priority': priority,
                'status': status,
            } });
    };
    /**
     * Get rights
     * Returns a list of rights.
     * @returns BakingRight
     * @throws ApiError
     */
    RightsService.prototype.get = function (_a) {
        var type = _a.type, baker = _a.baker, cycle = _a.cycle, level = _a.level, slots = _a.slots, priority = _a.priority, status = _a.status, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __awaiter(this, void 0, void 0, function () {
            var options, result;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        options = this.getApiRequestOptions({
                            type: type,
                            baker: baker,
                            cycle: cycle,
                            level: level,
                            slots: slots,
                            priority: priority,
                            status: status,
                            select: select,
                            sort: sort,
                            offset: offset,
                            limit: limit,
                        });
                        return [4 /*yield*/, __request(options)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, result.body];
                }
            });
        });
    };
    /**
     * **used to get the request options without making a http request**
     * Get rights
     * Returns a list of rights.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    RightsService.prototype.getApiRequestOptions = function (_a) {
        var type = _a.type, baker = _a.baker, cycle = _a.cycle, level = _a.level, slots = _a.slots, priority = _a.priority, status = _a.status, select = _a.select, sort = _a.sort, offset = _a.offset, _b = _a.limit, limit = _b === void 0 ? 100 : _b;
        return __assign(__assign({}, this.config), { method: 'GET', path: "/v1/rights", query: {
                'type': type,
                'baker': baker,
                'cycle': cycle,
                'level': level,
                'slots': slots,
                'priority': priority,
                'status': status,
                'select': select,
                'sort': sort,
                'offset': offset,
                'limit': limit,
            } });
    };
    return RightsService;
}());
export { RightsService };
