// @taqueria/protocol/ParsedPluginInfo.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// ParsedPluginInfo.ts
var ParsedPluginInfo_exports = {};
__export(ParsedPluginInfo_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema2,
  schemas: () => schemas2
});
module.exports = __toCommonJS(ParsedPluginInfo_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var ParsedOperation = __toESM(require("@taqueria/protocol/ParsedOperation"));
var PluginInfo = __toESM(require("@taqueria/protocol/PluginInfo"));
var import_zod = require("zod");
var internalSchema2 = PluginInfo.internalSchema.extend({
  operations: import_zod.z.preprocess(
    (val) => val != null ? val : [],
    import_zod.z.array(
      ParsedOperation.schemas.schema,
      { description: "ParsedOperations" }
    ).optional()
  )
}).describe("ParsedPluginInfo");
var rawSchema2 = PluginInfo.internalSchema.extend({
  operations: import_zod.z.preprocess(
    (val) => val != null ? val : [],
    import_zod.z.array(
      ParsedOperation.rawSchema,
      { description: "ParsedOperation" }
    )
  ).optional()
}).describe("ParsedPluginInfo");
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  rawSchema: rawSchema2,
  internalSchema: internalSchema2,
  parseErrMsg: (value) => `The following plugin info gave us trouble when parsing the following plugin information: ${value}`,
  unknownErrMsg: "Something went wrong trying to parse the plugin information"
});
var { create, of, make } = factory;
var schemas2 = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
