// @taqueria/protocol/SanitizedAbsPath.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// SanitizedAbsPath.ts
var SanitizedAbsPath_exports = {};
__export(SanitizedAbsPath_exports, {
  create: () => create,
  factory: () => factory,
  generatedSchemas: () => generatedSchemas,
  make: () => make,
  of: () => of,
  rawSchema: () => rawSchema,
  schemas: () => schemas
});
module.exports = __toCommonJS(SanitizedAbsPath_exports);
var import_Base = __toESM(require("@taqueria/protocol/Base"));
var path = __toESM(require("path"));
var import_zod = require("zod");
var rawSchema = import_zod.z.string({ description: "SanitizedAbsPath" }).min(1);
var { schemas: generatedSchemas, factory } = (0, import_Base.default)({
  isStringLike: true,
  rawSchema,
  transformer: (value) => path.resolve(value),
  parseErrMsg: (value) => `${value} is an invalid absolute path`,
  unknownErrMsg: (value) => `Something went wrong trying to parse the absolute path, ${value}`
});
var { create, make, of } = factory;
var schemas = {
  ...generatedSchemas,
  schema: generatedSchemas.schema.transform((val) => val)
};
