// @taqueria/protocol/taqueria-protocol-types.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// taqueria-protocol-types.ts
var taqueria_protocol_types_exports = {};
__export(taqueria_protocol_types_exports, {
  Alias: () => Alias,
  Config: () => Config,
  EconomicalProtocolHash: () => EconomicalProtocolHash,
  Environment: () => Environment,
  EphemeralState: () => EphemeralState,
  HumanReadableIdentifier: () => HumanReadableIdentifier,
  InstalledPlugin: () => InstalledPlugin,
  LoadedConfig: () => LoadedConfig,
  MetadataConfig: () => MetadataConfig,
  NetworkConfig: () => NetworkConfig,
  Operation: () => Operation,
  Option: () => Option,
  ParsedOperation: () => ParsedOperation,
  ParsedTemplate: () => ParsedTemplate,
  PersistentState: () => PersistentState,
  PluginActionName: () => PluginActionName,
  PluginInfo: () => PluginInfo,
  PluginJsonResponse: () => PluginJsonResponse,
  PluginResponseEncoding: () => PluginResponseEncoding,
  PluginSchema: () => PluginSchema,
  PositionalArg: () => PositionalArg,
  RequestArgs: () => RequestArgs,
  SHA256: () => SHA256,
  SandboxAccountConfig: () => SandboxAccountConfig,
  SandboxConfig: () => SandboxConfig,
  SanitizedAbsPath: () => SanitizedAbsPath,
  SanitizedArgs: () => SanitizedArgs,
  SanitizedPath: () => SanitizedPath,
  ScaffoldConfig: () => ScaffoldConfig,
  TaqError: () => TaqError,
  Task: () => Task,
  Template: () => Template,
  Url: () => Url,
  Verb: () => Verb,
  VersionNumber: () => VersionNumber,
  i18n: () => i18n
});
module.exports = __toCommonJS(taqueria_protocol_types_exports);
var Alias = __toESM(require("@taqueria/protocol/Alias"));
var Config = __toESM(require("@taqueria/protocol/Config"));
var EconomicalProtocolHash = __toESM(require("@taqueria/protocol/EconomicalProtocolHash"));
var Environment = __toESM(require("@taqueria/protocol/Environment"));
var EphemeralState = __toESM(require("@taqueria/protocol/EphemeralState"));
var HumanReadableIdentifier = __toESM(require("@taqueria/protocol/HumanReadableIdentifier"));
var i18n = __toESM(require("@taqueria/protocol/i18n"));
var InstalledPlugin = __toESM(require("@taqueria/protocol/InstalledPlugin"));
var LoadedConfig = __toESM(require("@taqueria/protocol/LoadedConfig"));
var MetadataConfig = __toESM(require("@taqueria/protocol/MetadataConfig"));
var NetworkConfig = __toESM(require("@taqueria/protocol/NetworkConfig"));
var Operation = __toESM(require("@taqueria/protocol/Operation"));
var Option = __toESM(require("@taqueria/protocol/Option"));
var ParsedOperation = __toESM(require("@taqueria/protocol/ParsedOperation"));
var ParsedTemplate = __toESM(require("@taqueria/protocol/ParsedTemplate"));
var PersistentState = __toESM(require("@taqueria/protocol/PersistentState"));
var PluginActionName = __toESM(require("@taqueria/protocol/PluginActionName"));
var PluginInfo = __toESM(require("@taqueria/protocol/PluginInfo"));
var PluginJsonResponse = __toESM(require("@taqueria/protocol/PluginJsonResponse"));
var PluginResponseEncoding = __toESM(require("@taqueria/protocol/PluginResponseEncoding"));
var PluginSchema = __toESM(require("@taqueria/protocol/PluginSchema"));
var PositionalArg = __toESM(require("@taqueria/protocol/PositionalArg"));
var RequestArgs = __toESM(require("@taqueria/protocol/RequestArgs"));
var SandboxAccountConfig = __toESM(require("@taqueria/protocol/SandboxAccountConfig"));
var SandboxConfig = __toESM(require("@taqueria/protocol/SandboxConfig"));
var SanitizedAbsPath = __toESM(require("@taqueria/protocol/SanitizedAbsPath"));
var SanitizedArgs = __toESM(require("@taqueria/protocol/SanitizedArgs"));
var SanitizedPath = __toESM(require("@taqueria/protocol/SanitizedPath"));
var ScaffoldConfig = __toESM(require("@taqueria/protocol/ScaffoldConfig"));
var SHA256 = __toESM(require("@taqueria/protocol/SHA256"));
var TaqError = __toESM(require("@taqueria/protocol/TaqError"));
var Task = __toESM(require("@taqueria/protocol/Task"));
var Template = __toESM(require("@taqueria/protocol/Template"));
var Url = __toESM(require("@taqueria/protocol/Url"));
var Verb = __toESM(require("@taqueria/protocol/Verb"));
var VersionNumber = __toESM(require("@taqueria/protocol/VersionNumber"));
