// @dipdup/tzkt-api/src/services/CyclesService.ts
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
import type { Cycle } from '../models/Cycle';
import type { Int32Parameter } from '../models/Int32Parameter';
import type { OffsetParameter } from '../models/OffsetParameter';
import type { SelectParameter } from '../models/SelectParameter';
import type { SortParameter } from '../models/SortParameter';
import type { Symbols } from '../models/Symbols';
import { ApiRequestOptions } from '../core/ApiRequestOptions';
import { request as __request } from '../core/request';
import { ApiOptions } from '../core/ApiOptions';

export class CyclesService {

    private readonly config: ApiOptions;

    /**
     * creates a new service
     * @param config the configuration to use
     */
    public constructor(config: ApiOptions) {
        this.config = config;
    }

    /**
     * Get cycles count
     * Returns the total number of cycles, including future cycles.
     * @returns number
     * @throws ApiError
     */
    public async getCount(): Promise<number> {
        const options = this.getCountApiRequestOptions(        );
        const result = await __request(options);
        return result.body;
    }

    /**
     * **used to get the request options without making a http request**
     * Get cycles count
     * Returns the total number of cycles, including future cycles.
     * @returns ApiRequestOptions the request options to fulfill a http request
     * @throws ApiError
     */
    public getCountApiRequestOptions(): ApiRequestOptions {
        return {
            ...this.config,
            method: 'GET',
            path: `/v1/cycles/count`,
        };
    }

    /**
     * Get cycles
     * Returns a list of cycles, including future cycles.
     * @returns Cycle
     * @throws ApiError
     */
    public async get({
        quote,
        snapshotIndex,
        select,
        sort,
        offset,
        limit = 100,
    }: {
        /** Comma-separated list of ticker symbols to inject historical prices into response **/
        quote: Symbols,
        /** Filters cycles by snapshot index (0..15) **/
        snapshotIndex?: Int32Parameter | null,
        /** Specify comma-separated list of fields to include into response or leave it undefined to return full object. If you select single field, response will be an array of values in both `.fields` and `.values` modes. **/
        select?: SelectParameter | null,
        /** Sorts cycles by specified field. Supported fields: `index` (default, desc). **/
        sort?: SortParameter | null,
        /** Specifies which or how many items should be skipped **/
        offset?: OffsetParameter | null,
        /** Maximum number of items to return **/
        limit?: number,
    }): Promise<Array<Cycle>> {
        const options = this.getApiRequestOptions({
            quote: quote,
            snapshotIndex: snapshotIndex,
            select: select,
            sort: sort,
            offset: offset,
            limit: limit,
        }
    );
    const result = await __request(options);
    return result.body;
}

/**
 * **used to get the request options without making a http request**
 * Get cycles
 * Returns a list of cycles, including future cycles.
 * @returns ApiRequestOptions the request options to fulfill a http request
 * @throws ApiError
 */
public getApiRequestOptions({
    quote,
    snapshotIndex,
    select,
    sort,
    offset,
    limit = 100,
}: {
    /** Comma-separated list of ticker symbols to inject historical prices into response **/
    quote: Symbols,
    /** Filters cycles by snapshot index (0..15) **/
    snapshotIndex?: Int32Parameter | null,
    /** Specify comma-separated list of fields to include into response or leave it undefined to return full object. If you select single field, response will be an array of values in both `.fields` and `.values` modes. **/
    select?: SelectParameter | null,
    /** Sorts cycles by specified field. Supported fields: `index` (default, desc). **/
    sort?: SortParameter | null,
    /** Specifies which or how many items should be skipped **/
    offset?: OffsetParameter | null,
    /** Maximum number of items to return **/
    limit?: number,
}): ApiRequestOptions {
    return {
        ...this.config,
        method: 'GET',
        path: `/v1/cycles`,
        query: {
            'quote': quote,
            'snapshotIndex': snapshotIndex,
            'select': select,
            'sort': sort,
            'offset': offset,
            'limit': limit,
        },
    };
}

/**
 * Get cycle by index
 * Returns a cycle at the specified index.
 * @returns Cycle
 * @throws ApiError
 */
public async getByIndex({
    index,
    quote,
}: {
    /** Cycle index starting from zero **/
    index: number,
    /** Comma-separated list of ticker symbols to inject historical prices into response **/
    quote: Symbols,
}): Promise<Cycle> {
    const options = this.getByIndexApiRequestOptions({
        index: index,
        quote: quote,
    }
);
const result = await __request(options);
return result.body;
}

/**
 * **used to get the request options without making a http request**
 * Get cycle by index
 * Returns a cycle at the specified index.
 * @returns ApiRequestOptions the request options to fulfill a http request
 * @throws ApiError
 */
public getByIndexApiRequestOptions({
    index,
    quote,
}: {
    /** Cycle index starting from zero **/
    index: number,
    /** Comma-separated list of ticker symbols to inject historical prices into response **/
    quote: Symbols,
}): ApiRequestOptions {
    return {
        ...this.config,
        method: 'GET',
        path: `/v1/cycles/${index}`,
        query: {
            'quote': quote,
        },
    };
}

}
